### PR TITLE
Implement DATE_FORMAT function

### PR DESCRIPTION
**Pull request by didendron**
Implement DATE_FORMAT according to the MySQL spec: https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-format 
Tested against the actual output from MySQL using randomized inputs
Documentation
Integration tests

Relates to https://github.com/elastic/elasticsearch/issues/55065

### CODE REVIEW

**Comment by didendron** [comment]
@cla-bot check

**Comment by luigidellaquila** [comment]
@elasticmachine test this please

File: docs/reference/sql/functions/date-time.asciidoc
Comment (luigidellaquila) [inline docs/reference/sql/functions/date-time.asciidoc:423]: For consistency, I'd move this section right after TO_CHAR, so that we have all the formatting functions together (SqlServer, PostgreSQL, MySQL)

File: x-pack/plugin/sql/qa/server/src/main/resources/datetime.csv-spec
Comment (luigidellaquila) [inline x-pack/plugin/sql/qa/server/src/main/resources/datetime.csv-spec:655]: It would be nice to have some tests that cover patterns that contain random strings, eg. 

```
DATE_FORMAT(some_date, 'he was born on the %dth of %M')
```

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatTestScript.java
Comment (luigidellaquila) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatTestScript.java:26]: I like the idea of having randomized tests and to check it against a real MySQL instance.
Though, IMHO this class needs a better name (eg. `DateFormatRandomDatasetGenerator`) and a comment that explains its purpose and that points at the tests that are using its output.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeDateFormatProcessorTests.java
Comment (luigidellaquila) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeDateFormatProcessorTests.java:40]: I think this Javadoc is invalid, ie. `<` and `>` should be `&lt;` and `&gt;`

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java
Comment (luigidellaquila) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java:159]: This class is private, these constructor/methods should not be public

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatTestScript.java
Comment (luigidellaquila) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatTestScript.java:41]: Please, use `org.elasticsearch.common.Randomness.get()` here

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeDateFormatProcessorTests.java
Comment (luigidellaquila) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeDateFormatProcessorTests.java:88]: Please, use `DateTimeFormatter.ofPattern(String, Locale)` here

**Review by luigidellaquila — CHANGES_REQUESTED** [review]
Hi @didendron

First of all, thank you very much for this contribution. I had a quick look at it and it's clear that you put some significant effort and attention on it.

Before we can proceed and merge it, we have to make sure it matches Elasticsearch quality standards, so please bear with me in this review process.

As a first step, I have to ask you to do the following:

1. add a changelog file `/docs/changelog/88388.yaml` with the following content:
```
pr: 88388
summary: Implement DATE_FORMAT function
area: SQL
type: enhancement
issues:
 - 55065
```

2. run the tests locally (start with ` ./gradlew precommit` and `./gradlew -p x-pack/plugin/sql --parallel check`), you will see some errors and compliance problems; we cannot merge the PR as long as they fail

3. I added a first round comments to the code, please have a look at them and see if they can be addressed

If you need any clarification, please let me know

Thanks

Luigi

File: docs/reference/sql/functions/date-time.asciidoc
Comment (luigidellaquila) [reply docs/reference/sql/functions/date-time.asciidoc:423]: Well, the functions should be in alphabetical order actually, so this should be fine.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatTestScript.java
Comment (didendron) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatTestScript.java]: I tried to use the method org. elasticsearch. common.Randomness. get(), but it caused an error,
so I used a solution similar to the one in ToCharTestScript.java

**Comment by luigidellaquila** [comment]
Hi @didendron 

Thank you very much for the quick feedback on my comments.
I ran the tests locally and I see `ConsistentFunctionArgHandlingIT` is still failing. I think adding the function name to `ConsistentFunctionArgHandlingIT-non-tested-functions.txt` should be enough to fix it.

After that, could you please run the tests again and see if something else is missing?

Thanks

Luigi

**Comment by luigidellaquila** [comment]
@elasticmachine test this please

**Comment by luigidellaquila** [comment]
Hi @didendron 

Thank you very much, this looks good for me now.
Please, just align the PR with `main` branch (ie. merge main branch) so that I can run some final tests.

Thanks!

**Comment by luigidellaquila** [comment]
@elasticmachine test this please

**Review by luigidellaquila — APPROVED** [review]

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java:154]: This class can also be made entirely `private` (including all its methods).

**Review by astefan — COMMENTED** [review]
Left few minor comments. Looking good so far.

File: x-pack/plugin/sql/qa/server/src/main/resources/datetime.csv-spec
Comment (didendron) [reply x-pack/plugin/sql/qa/server/src/main/resources/datetime.csv-spec:653]: ok

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java
Comment (didendron) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java]: ok

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java
Comment (didendron) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java]: ok

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java:103]: I had a different change in mind: **no** access modifiers, meaning the default one ("package protected").

**Review by astefan — APPROVED** [review]
LGTM. Left a follow up comment regarding the access modifier used in DateFormatter.

**Review by Luegg — APPROVED** [review]
Lgtm

**Comment by luigidellaquila** [comment]
Hi @didendron 

Thank you very much for following up on this, I think we are almost there.
If you could just address last comment from @astefan, I'd go on and merge

Thanks

Luigi

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java
Comment (didendron) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java:103]: Yes of course

**Comment by didendron** [comment]
@Luegg, @astefan, and @luigidellaquila in particular thank you for your help and support.

**Comment by astefan** [comment]
Hi @didendron,
Please, align the PR with main branch (ie. merge main branch) so that we can run the tests one more time with your new updates. Thanks.

**Comment by astefan** [comment]
@elasticmachine test this please

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/docs

**Comment by luigidellaquila** [comment]
Thank you very much @didendron, I'm merging the PR

**Pull request merged**
